File: src/main/java/ru/citeck/ecos/index/IndexKey.java
package ru.citeck.ecos.index;

import com.intellij.util.io.KeyDescriptor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.jetbrains.annotations.NotNull;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import java.util.Objects;

@Data
@EqualsAndHashCode
public class IndexKey {

    public static final String SEARCH_EVERYWHERE_TYPE = "search-everywhere";
    public static final String REFERENCE_TYPE = "reference";
    public static final String ALL = "*";
    public static final IndexKey SEARCH_EVERYWHERE = new IndexKey(SEARCH_EVERYWHERE_TYPE, ALL);
    private final String type;
    private final String id;
    public IndexKey(String type) {
        this.type = type;
        this.id = ALL;
    }

    public IndexKey(String type, String id) {
        this.type = type;
        this.id = id;
    }

    public static class Descriptor implements KeyDescriptor<IndexKey> {

        @Override
        public int getHashCode(IndexKey key) {
            return key.hashCode();
        }

        @Override
        public boolean isEqual(IndexKey key1, IndexKey key2) {
            return Objects.equals(key1, key2);
        }

        @Override
        public void save(@NotNull DataOutput out, IndexKey key) throws IOException {
            out.writeUTF(key.getType());
            out.writeUTF(key.getId());
        }

        @Override
        public IndexKey read(@NotNull DataInput in) throws IOException {
            return new IndexKey(
                    in.readUTF(),
                    in.readUTF()
            );
        }

    }

}

File: src/main/kotlin/ru/citeck/idea/search/index/IndexKey.kt
package ru.citeck.idea.search.index

import com.intellij.util.io.KeyDescriptor
import java.io.DataInput
import java.io.DataOutput

data class IndexKey(
    val type: String,
    val id: String
) {

    companion object {
        const val SEARCH_EVERYWHERE_TYPE: String = "search-everywhere"
        const val REFERENCE_TYPE: String = "reference"
        const val ARTIFACT_TYPE: String = "artifact-type"
        const val ALL: String = "*"
        val SEARCH_EVERYWHERE: IndexKey = IndexKey(SEARCH_EVERYWHERE_TYPE, ALL)
    }

    constructor(type: String) : this(type, ALL)

    class Descriptor : KeyDescriptor<IndexKey> {
        override fun getHashCode(key: IndexKey): Int {
            return key.hashCode()
        }

        override fun isEqual(key1: IndexKey, key2: IndexKey): Boolean {
            return key1 == key2
        }

        override fun save(output: DataOutput, key: IndexKey) {
            output.writeUTF(key.type)
            output.writeUTF(key.id)
        }

        override fun read(input: DataInput): IndexKey {
            return IndexKey(
                input.readUTF(),
                input.readUTF()
            )
        }
    }
}
